src/template/wasm2c.declarations.c

```diff
@@ -15,10 +15,19 @@
 
 #define UNREACHABLE TRAP(UNREACHABLE)
 
-#define CALL_INDIRECT(table, t, ft, x, ...)             \
-  (LIKELY((x) < table.size && table.data[x].func &&     \
-          table.data[x].func_type == func_types[ft]) || \
-       TRAP(CALL_INDIRECT),                             \
+static inline bool func_types_eq(wasm_rt_func_type_t a, wasm_rt_func_type_t b) {
+  if (UNLIKELY(a.len != b.len)) {
```

Comment:
Drive-by comment, feel free to disregard: if you guaranteed that the `uint64_t` and `const char*` representations never collided (e.g. by always aligning the string allocations to an even byte boundary and setting the LSB of the uint64_t encoding to 1) then const char* equality implies string equality as well, which let's you remove two branches in the common case and just do:

if (a.type.val == b.type.val) {
  return true;
}
if (a.type.val & 1 || a.type.val & 1) {
  return false;
}
if (a.type.len != b.type.len) {
  return false;
}
return !memcmp(a.type.str, b.type.str, a.len);

Reply:
Hmm, thanks! I think only **one** of those branches is there today; the "a" argument (including a.len) is always known at compile time. But I think yours saves another branch.

It may be a little tricky to figure out how to arrange this:
- I don't know how to guarantee how the linker aligns a static char* string literal (do you?).
- Right now, every bit in the uint64_t is spoken for.
- What are the rules on the legality of comparing `a.type.val == b.type.val` if "str" is the active member of the union (and if the two members aren't the same length, e.g. on a 32-bit system)? How do we make this safe?

Reply:
Re alignment: With C11 you have _Alignas(2), but maybe we have to support further back than that. One option might be to simply use "static const uint16_t[]" instead of "static const char[]".

On that note, you could maybe also use \0 termination in the array to save you from having to also store length (saves a bit on cache/memory pressure, maybe).

In terms of validity, this sort of thing is fine when operating over trivial/POD types (you would want to make sure your pointers are 64-bit though and pad them out if not -- maybe use uint64_t regardless to store the value and do a pointer cast when using as a pointer instead). Those are the rules for C++ so I imagine C is also fine with it.

That said, if all bits are spoken for then disregard. I just thought I'd offer this up as I'm starting to look into using wasm2c and indirect call performance will be important.

Reply:
Hmm, there may be a much simpler way to do this. What if the functype is a pointer (perhaps to a {length,const char*}, or maybe to four uint64_t's that are the output of a SHA-256), and the condition for equality is that *either* the pointers are identical, or the target of the pointers are identical? And then each module can canonicalize its own function types so that identical types get an identical pointer.

That way, a call_indirect within the module will be a single pointer comparison, end of story. And a call_indirect for a funcref imported from another module will generally involve comparing four uint64_t's (or a memcmp between two length-tagged memory regions).

How would you feel about that... is it enough for "same-module" call_indirects to be extremely fast, and "other-module" call-indirects to be, I dunno, still extremely fast but probably 2 cycles or something?

Reply:
Sounds good to me! Much cleaner. Thank you!

Reply:
Ok done, and thanks for the idea! I think this is a lot simpler, and because of (string) constant merging in the GNU/LLVM toolchains, it basically doesn't affect performance compared with the main branch because pointer equality is almost always the end of the story. Also there's no difference anymore between "large-signature" functions and short-signature functions.